include/secp256k1.h

```diff
@@ -229,7 +231,7 @@ SECP256K1_API secp256k1_context* secp256k1_context_clone(
  *  be used instead.
  *
  *  Args:   ctx: an existing context to destroy, constructed using
- *               secp256k1_context_create or secp256k1_context_clone
+ *               secp256k1_context_create or secp256k1_context_clone.
  */
 SECP256K1_API void secp256k1_context_destroy(
```

Comment:
Does `secp256k1_context_destroy`'s `ctx` deserves its own `SECP256K1_ARG_NONNULL` annotation?

edit: Have a look on `secp256k1_ecdsa_signature_normalize`'s second arg, missing too?
re-edit: `secp256k1_context_randomize`'s second arg?

Reply:
Good question if it shouldn't be null, right now it will work if it's not but we don't promise that so I think it should have.

As for the rest:
`secp256k1_ecdsa_signature_normalize` 2nd argument can be NULL if you just want to check for normalization without normalizing.
`secp256k1_context_randomize` you can pass NULL as the 2nd argument to reset to the initial state.

Reply:
I don't think it's worth changing the API of `context_destroy` and `preallocated_destroy` given that it is fine to give NULL pointers to `free()`.

Reply:
Good point, though i don't know how "if ptr NULL, no operations is performed" guarantee holds beyond standardized *nix platforms.

Reply:
Probably best to mention this behavior in the doc ("Ignored if NULL" or something).

Reply:
@jonasnick See https://github.com/bitcoin-core/secp256k1/pull/783#discussion_r462912364

Reply:
If we don't mention this, we will have the exact same discussion again at some point. But fine either way imo.

Reply:
@ariard that behavior is part of the C standard (http://port70.net/~nsz/c/c89/c89-draft.html#4.10.3.2)

Reply:
Yeah, I wanted to make that promise too, but not too important (as long as it's explicitly either NON_NULL or allowed to be null, I prefer not to leave it in a limbo)

Reply:
> @ariard that behavior is part of the C standard (http://port70.net/~nsz/c/c89/c89-draft.html#4.10.3.2)

Ah right, so "allowed to be null" recalling C standard?

Reply:
@ariard Not sure if we're talking past each other. Sure the current `free(NULL)` implementation is fine. I was just pointing out that adding the promise that `secp256k1_context_destroy(NULL)` is a no-op will be an API change and I don't see a reason to make this change.

Reply:
I still think it would be better to remove the `SECP256K1_ARG_NONNULL` because it is a backwards-incompatible API change.

Reply:
Just to clarify, it's an API change since before 0881633d we didn't force that pointer args must not be NULL if not mentioned otherwise. But here we never mentioned "cannot be NULL" or similar.